samples/subsys/fs/zms/src/main.c

```diff
@@ -0,0 +1,282 @@
+ * 4. A long set of data (128 bytes)
+ *
+ * A loop is executed where we mount the storage system, and then write all set
+ * of data.
+ *
+ * Each DELETE_ITERATION period, we delete all set of data and verify that it has been deleted.
+ * We generate as well incremented ID/value pairs, we store them until storage is full, then we
+ * delete them and verify that storage is empty.
+ *
+ * Copyright (c) 2024 BayLibre SAS
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/sys/reboot.h>
+#include <zephyr/device.h>
+#include <string.h>
+#include <zephyr/drivers/flash.h>
+#include <zephyr/storage/flash_map.h>
+#include <zephyr/fs/zms.h>
+
+static struct zms_fs fs;
+
+#define ZMS_PARTITION        storage_partition
+#define ZMS_PARTITION_DEVICE FIXED_PARTITION_DEVICE(ZMS_PARTITION)
+#define ZMS_PARTITION_OFFSET FIXED_PARTITION_OFFSET(ZMS_PARTITION)
+
+#define IP_ADDRESS_ID 1
+#define KEY_VALUE_ID  0xbeefdead
+#define CNT_ID        2
+#define LONG_DATA_ID  3
+
+#define MAX_ITERATIONS   300
+#define DELETE_ITERATION 10
+
+static int delete_and_verify_items(struct zms_fs *fs, uint32_t id)
+{
+	int rc = 0;
+
+	rc = zms_delete(fs, id);
+	if (rc) {
+		goto error1;
+	}
+	rc = zms_get_data_length(fs, id);
+	if (rc > 0) {
+		goto error2;
+	}
+
+	return 0;
+error1:
+	printk("Error while deleting item rc=%d\n", rc);
+	return rc;
+error2:
+	printk("Error, Delete failed item should not be present\n");
+	return -1;
+}
+
+static int delete_basic_items(struct zms_fs *fs)
+{
+	int rc = 0;
+
+	rc = delete_and_verify_items(fs, IP_ADDRESS_ID);
+	if (rc) {
+		printk("Error while deleting item %x rc=%d\n", IP_ADDRESS_ID, rc);
+		return rc;
+	}
+	rc = delete_and_verify_items(fs, KEY_VALUE_ID);
+	if (rc) {
+		printk("Error while deleting item %x rc=%d\n", KEY_VALUE_ID, rc);
+		return rc;
+	}
+	rc = delete_and_verify_items(fs, CNT_ID);
+	if (rc) {
+		printk("Error while deleting item %x rc=%d\n", CNT_ID, rc);
+		return rc;
+	}
+	rc = delete_and_verify_items(fs, LONG_DATA_ID);
+	if (rc) {
+		printk("Error while deleting item %x rc=%d\n", LONG_DATA_ID, rc);
+	}
+
+	return rc;
+}
+
+int main(void)
+{
+	int rc = 0;
+	char buf[16];
+	uint8_t key[8] = {0xDE, 0xAD, 0xBE, 0xEF, 0xDE, 0xAD, 0xBE, 0xEF}, longarray[128];
+	uint32_t i_cnt = 0U, i;
+	uint32_t id = 0;
+	ssize_t free_space = 0;
+	struct flash_pages_info info;
+
+	for (int n = 0; n < sizeof(longarray); n++) {
+		longarray[n] = n;
+	}
+
+	/* define the zms file system by settings with:
+	 *	sector_size equal to the pagesize,
+	 *	3 sectors
+	 *	starting at ZMS_PARTITION_OFFSET
+	 */
+	fs.flash_device = ZMS_PARTITION_DEVICE;
+	if (!device_is_ready(fs.flash_device)) {
+		printk("Storage device %s is not ready\n", fs.flash_device->name);
+		return 0;
+	}
+	fs.offset = ZMS_PARTITION_OFFSET;
+	rc = flash_get_page_info_by_offs(fs.flash_device, fs.offset, &info);
+	if (rc) {
+		printk("Unable to get page info, rc=%d\n", rc);
+		return 0;
+	}
+	fs.sector_size = info.size;
```

Comment:
@andrisk-dev yes I think that we should be able to set a custom sector size instead of retrieving this information form the flash page info (which is the size of the erase block)

Reply:
> he way how the sector_size is decided in the example works in most cases, but will cause problems in case when flash write_block_size is the same (or not much smaller) as erase_block_size.

I added a Config option in the last revision to customize sector counts

Reply:
Maybe I overlooked something but it seems to me that fs.sector_size is still not configurable.

Reply:
Hi @andrisk-dev, I actually removed this from the last revision as it creates some confusion.
ZMS should receive the sector_size from upper layers and should not modify it locally.
If you use the ZMS directly you should set the sector_size according to your setup (partition size, write-block-size, ...)
In my sample App I set it using flash_layout info which is the page size : fs.sector_size = info.size;
But this value could also be set using the partition size which you can get through DT_REG_SIZE macro and adjust it according to the sector count.
So in general upper layers should modify the sector size and not the other way around

Reply:
@andrisk-dev As an example, if you want to use ZMS through Settings subsystem, you can just modify this config to adjust the sector_size : [here](https://github.com/zephyrproject-rtos/zephyr/pull/78632/commits/b99b214b949d3ae10bf02eb60f4ade4fb42400a1#diff-1d8e03c3de8c34f0dadc39418111e718253db43ffa285c36e5c361ffd02cf45bR391)

Reply:
hi @rghaddab. I agree that upper layers should set it. But I thought about some setting for the example itself. I think I should not be changing info.size to change the behavior of the example for one particular development board. I am not sure what is the best place to change the behavior of the ZMS example but I don't see a way how to easily configure the example for the case I described in the original comment.

Reply:
@andrisk-dev I get your point.
Currently the first version of the sample code supports the native_sim target that's why it gets the sector size information from info.size.
To support boards like what you described, in the sample code, this part should be changed to set the sector size based on the storage partition size.
I propose to add this after this first sample get merged, and you can help to do that, what do you think ?